### PR TITLE
Exhaustive recovery

### PR DESCRIPTION
**Pull request by apoelstra**
Builds on #426. Bumps branch coverage for modules/recovery/main_impl.h to 100%.

### CODE REVIEW

File: src/tests_exhaustive.c [hunk 1]
```diff
@@ -282,6 +287,130 @@ void test_exhaustive_sign(const secp256k1_context *ctx, const secp256k1_ge *grou
...
+                    /* nb we have a "high s" rule */
+                    should_verify &= !secp256k1_scalar_is_high(&s_s);
+
+                    /* We would like to try recovering the pubkey and checking that it matches,
+                     * but pubkey recovery is impossible in the exhaustive tests (the reason
+                     * being that there are 12 nonzero r values, 12 nonzero points, and no
+                     * overlap between the sets, so there are no valid signatures). */
+
+                    /* Verify by converting to a standard signature and calling verify */
+                    secp256k1_ecdsa_recoverable_signature_save(&rsig, &r_s, &s_s, recid);
```
Comment (gmaxwell) [inline src/tests_exhaustive.c]: recid is used with initialization here.

**Comment by apoelstra** [comment]
Fixed uninitialized `recid` and also destroyed `ctx` so that valgrind would run clean on exhaustive tests.

**Comment by gmaxwell** [comment]
ACK.

**Comment by sipa** [comment]
ACK

**Pull request merged**